src/main/java/com/google/cloud/flink/bigquery/ProtobufUtils.java

```diff
@@ -0,0 +1,717 @@
+    } catch (Descriptors.DescriptorValidationException e) {
+      throw new IllegalArgumentException("Could not build Proto-Schema from Spark schema", e);
+    }
+  }
+
+  public static ProtoSchema toProtoSchema(RowType schema) throws IllegalArgumentException {
+    try {
+      Descriptors.Descriptor descriptor = toDescriptor(schema);
+      return ProtoSchemaConverter.convert(descriptor);
+    } catch (Descriptors.DescriptorValidationException e) {
+      throw new IllegalArgumentException("Could not build Proto-Schema from Spark schema", e);
+    }
+  }
+
+  public static ProtoSchema toProtoSchema(StructuredType schema) throws IllegalArgumentException {
+    try {
+      Descriptors.Descriptor descriptor = toDescriptor(schema);
+      return ProtoSchemaConverter.convert(descriptor);
+    } catch (Descriptors.DescriptorValidationException e) {
+      throw new IllegalArgumentException("Could not build Proto-Schema from Spark schema", e);
+    }
+  }
+
+  private static DescriptorProtos.FieldDescriptorProto.Type toProtoFieldType(
+      LegacySQLTypeName bqType) {
+    if (LegacySQLTypeName.RECORD.equals(bqType)) {
+      throw new IllegalStateException(
+          "Program attempted to return an atomic data-type for a RECORD");
+    }
+    return Preconditions.checkNotNull(
+        BigQueryToProtoType.get(bqType),
+        new IllegalArgumentException("Unexpected type: " + bqType.name()));
+  }
+
+  public static Descriptors.Descriptor toDescriptor(RowType schema)
+      throws Descriptors.DescriptorValidationException {
+    DescriptorProtos.DescriptorProto.Builder descriptorBuilder =
+        DescriptorProtos.DescriptorProto.newBuilder().setName("Schema");
+    List<Field> listOfFields = new ArrayList<Field>();
+    Iterator<RowField> streamdata = schema.getFields().iterator();
+    while (streamdata.hasNext()) {
+      listOfFields.addAll(getFields(streamdata.next(), null));
+    }
+    int initialDepth = 0;
+    DescriptorProtos.DescriptorProto descriptorProto =
+        buildDescriptorProtoWithFields(descriptorBuilder, FieldList.of(listOfFields), initialDepth);
+
+    return createDescriptorFromProto(descriptorProto);
+  }
+
+  public static Descriptors.Descriptor toDescriptor(StructuredType schema)
+      throws Descriptors.DescriptorValidationException {
+    DescriptorProtos.DescriptorProto.Builder descriptorBuilder =
+        DescriptorProtos.DescriptorProto.newBuilder().setName("Schema");
+    ArrayList<Field> listOfFileds = new ArrayList<Field>();
+    Iterator<StructuredAttribute> streamdata = schema.getAttributes().stream().iterator();
+    while (streamdata.hasNext()) {
+      StructuredAttribute elem = streamdata.next();
+      if ("ROW".equals(elem.getType().getTypeRoot().toString())) {
+        Field.newBuilder(
+                elem.getName(),
+                StandardSQLTypeName.STRUCT,
+                FieldList.of(getListOfSubFields(elem.getType())))
+            .setMode(Mode.NULLABLE)
```

Comment:
Are Flink schemas are all nullable by default?

Reply:
Added condition if type is nullable then setting it as nullable otherwise mode as required .setMode(elem.getType().isNullable() ? Mode.NULLABLE : Mode.REQUIRED)